07-advanced-architecture/projects/final/app/src/main/java/com/kodeco/chat/conversation/Conversation.kt

```diff
@@ -0,0 +1,425 @@
+  }
+
+  val authorImageId: Int =
+    if (isUserMe) R.drawable.profile_photo_android_developer else R.drawable.someone_else
+  val spaceBetweenAuthors = if (isLastMessageByAuthor) Modifier.padding(top = 8.dp) else Modifier
+  Row(modifier = spaceBetweenAuthors) {
+    if (isLastMessageByAuthor) {
+      // Avatar
+      Image(
+        modifier = Modifier
+          .clickable(onClick = { onAuthorClick(msg.message.userId) })
+          .padding(horizontal = 16.dp)
+          .size(42.dp)
+          .border(1.5.dp, borderColor, CircleShape)
+          .border(3.dp, MaterialTheme.colorScheme.surface, CircleShape)
+          .clip(CircleShape)
+          .align(Alignment.Top),
+        painter = painterResource(id = authorImageId),
+        contentScale = ContentScale.Crop,
+        contentDescription = null
+      )
+    } else {
+      // Space under avatar
+      Spacer(modifier = Modifier.width(74.dp))
+    }
+    AuthorAndTextMessage(
+      msg = msg,
+      isUserMe = isUserMe,
+      isFirstMessageByAuthor = isFirstMessageByAuthor,
+      isLastMessageByAuthor = isLastMessageByAuthor,
+      authorClicked = onAuthorClick,
+      modifier = Modifier
+        .padding(end = 16.dp)
+        .weight(1f)
+    )
+  }
+}
+
+@Composable
+fun AuthorAndTextMessage(
+  msg: MessageUiModel,
+  isUserMe: Boolean,
+  isFirstMessageByAuthor: Boolean,
+  isLastMessageByAuthor: Boolean,
+  authorClicked: (String) -> Unit,
+  modifier: Modifier = Modifier
+) {
+  Column(modifier = modifier) {
+    if (isLastMessageByAuthor) {
+      AuthorNameTimestamp(msg, isUserMe)
+    }
+    ChatItemBubble(
+      msg.message,
+      isUserMe,
+      authorClicked = authorClicked
+    )
+    if (isFirstMessageByAuthor) {
+      // Last bubble before next author
+      Spacer(modifier = Modifier.height(8.dp))
+    } else {
+      // Between bubbles
+      Spacer(modifier = Modifier.height(4.dp))
+    }
+  }
+}
+
+@Composable
+private fun AuthorNameTimestamp(msg: MessageUiModel, isUserMe: Boolean = false) {
+  var userFullName: String = msg.user.fullName
+  if (isUserMe) {
+    userFullName = "me"
+  }
+
+  // Combine author and timestamp for author.
+  Row(modifier = Modifier.semantics(mergeDescendants = true) {}) {
+    Text(
+      text = userFullName,
+      style = MaterialTheme.typography.titleMedium,
+      modifier = Modifier
+        .alignBy(LastBaseline)
+        .paddingFrom(LastBaseline, after = 8.dp) // Space to 1st bubble
+    )
+    Spacer(modifier = Modifier.width(8.dp))
+    Text(
+      text = msg.message.createdOn.toString().isoToTimeAgo(),
+      style = MaterialTheme.typography.bodySmall,
+      modifier = Modifier.alignBy(LastBaseline),
+      color = MaterialTheme.colorScheme.onSurfaceVariant
+    )
+  }
+}
+
+
+@Composable
+fun ChatItemBubble(
+  message: Message,
+  isUserMe: Boolean,
+  authorClicked: (String) -> Unit
+) {
+  val chatBubbleShape = RoundedCornerShape(4.dp, 20.dp, 20.dp, 20.dp)
+  remember { mutableStateOf(false) }
```

Comment:
@abunur Is this necessary?

Reply:
Are you referencing the shape or the clickable message? The shape defines the shape for the chat bubble. The Clickable message composable handles click on the message. A message can contain a hyperlink - this is demonstrated in an earlier chapter - Chapter 5

Reply:
![dummy_chats_rendered](https://github.com/kodecocodes/aaf-materials/assets/4803881/c26cf4a0-e8b6-45ee-b958-becfe7f1eddc)

Reply:
I'm referring to the line:
```kotlin
remember { mutableStateOf(false) }
```
It's not assigned to anything.

So, is it necessary?

Reply:
It can be removed

Reply:
I removed it in main branch and pushed the changes, since I think these PRs are for @zlippard